File: desing-pattens-1/src/modelo/implementacao/desconto/DescontoOrcamentoMaiorQue500.java
package modelo.implementacao.desconto;

import interfaces.IDesconto;
import modelo.Orcamento;

public class DescontoOrcamentoMaiorQue500 implements IDesconto {

	private IDesconto proximo;

	@Override
	public double calcula(Orcamento orcamento) {

		if (orcamento.getValor() > 500) {
			return 0.07;
		}
		return proximo.calcula(orcamento);
	}

	@Override
	public void setProximo(IDesconto desconto) {

		this.proximo = desconto;
	}
}

File: desing-pattens-1/src/main/java/modelo/implementacao/desconto/DescontoOrcamentoMaiorQue500.kt
package main.java.modelo.implementacao.desconto

import main.java.interfaces.IDesconto
import main.java.modelo.Orcamento

class DescontoOrcamentoMaiorQue500 : IDesconto {

    private var proximo: IDesconto? = null

    override fun calcula(orcamento: Orcamento): Double {

        return if (orcamento.valor > 500) {
            0.07
        } else proximo!!.calcula(orcamento)
    }

    override fun setProximo(desconto: IDesconto) {

        this.proximo = desconto
    }
}
